OpenRA.Platforms.Default/OpenGL.cs

```diff
@@ -553,5 +662,56 @@ public static void WriteGraphicsLog(string message)
 			for (var i = 0; i < extensionCount; i++)
 				Log.Write("graphics", glGetStringi(GL_EXTENSIONS, (uint)i));
 		}
+
+		public static void CheckGLError()
+		{
+			// Let the debug message handler log the errors instead.
+			if (Features.HasFlag(GLFeatures.DebugMessagesCallback))
+				return;
+
+			var type = glGetError();
+			if (type == GL_NO_ERROR)
+				return;
+
+			var errorText = ErrorToText[type];
+			var error = "GL Error: {0}\n{1}".F(errorText, new StackTrace());
+
+			WriteGraphicsLog(error);
+
+			const string exceptionMessage = "OpenGL Error: See graphics.log for details.";
+
+			if (type == GL_OUT_OF_MEMORY)
+				throw new OutOfMemoryException(exceptionMessage);
+
+			throw new InvalidOperationException(exceptionMessage);
+		}
+
+		static void DebugMessageHandler(int source, int type, uint id, int severity, int length, StringBuilder message, IntPtr userparam)
+		{
+			string error;
+
+			switch (severity)
+			{
+				case GL_DEBUG_SEVERITY_HIGH:
+					error = BuildErrorText(source, type, severity, message);
+					WriteGraphicsLog(error);
+					throw new InvalidOperationException("OpenGL Error: See graphics.log for details.");
+
+				case GL_DEBUG_SEVERITY_MEDIUM:
+					error = BuildErrorText(source, type, severity, message);
+					Console.WriteLine(error);
+					break;
+			}
+		}
+
+		static string BuildErrorText(int source, int type, int severity, StringBuilder message)
+		{
+			string sourceText = DebugSourceToText[source];
+			string typeText = DebugTypeToText[type];
+			string severityText = DebugSeverityToText[severity];
```

Comment:
Are these lookups guaranteed to be exhaustive of all codes that might be thrown at them (same for `ErrorToText`)? Might be worth considering `TryGetValue`.